README.md
# readly-get
Permet de sauvegarder une copie d'une publication Readly.

Le but est de pouvoir lire une publication sur un support non compatible avec les applications fournies par Readly. 
Il est évident que les publications ne doivent en aucun cas être conservées une fois la lecture terminée ou lorsque votre abonnement ne vous permet plus de la lire. 
 
## readly_get.py
Permet de sauvegarder une copie d'une publication Readly. 
 
### Utilisation
```
usage: readly_get.py [-h] [--token TOKEN] [--output-folder OUTPUT_FOLDER] [--pattern PATTERN] [--image-format {jpeg,webp}] [--quality QUALITY] [--container-format {pdf,cbz}] [--low-quality] [--dpi DPI]
                     [--user-agent USER_AGENT] [--pause SECONDS] [--max-dl MAX_DL] [--no-clean] [--get-articles] [--get-articles-only] [--create-token] [--version]
                     [url]

Script to save a Readly publication.

positional arguments:
  url                   URL of the desired publication or publication_id.

options:
  -h, --help            show this help message and exit
  --token TOKEN         Authentication token or file containing the token. Default="auth_token".
  --output-folder OUTPUT_FOLDER, -o OUTPUT_FOLDER
                        Folder to save your publication. Default="DOWNLOADS".
  --pattern PATTERN     Pattern to name your file (available: "title", "issue", "date"). Default="title - issue (date)".
  --image-format {jpeg,webp}, -i {jpeg,webp}
                        Image format saved (available: "jpeg", "webp"). Default="jpeg".
  --quality QUALITY, -q QUALITY
                        Image quality (100 = best quality). Default="85".
  --container-format {pdf,cbz}, -c {pdf,cbz}
                        Output file type (available: "cbz", "pdf"). Default="pdf".
  --low-quality         Get default low quality images instead of HQ images.
  --dpi DPI             Image DPI (0 = original DPI). Default="300".
  --user-agent USER_AGENT
                        User-agent to use.
  --pause SECONDS, -p SECONDS
                        Make a pause (in seconds) between two pages. Default="0"
  --max-dl MAX_DL       Max number of issues to download. Default="1".
  --no-clean            Don't delete the temp folder where the images are stored.
  --get-articles        Also download attached articles. Use with "--no-clean" option, or files will be deleted.
  --get-articles-only   Download only attached articles (no image). Won't create PDF / CBZ file. Will force "--no-clean" option.
  --create-token        Create a new token.
  --version             Current version.
```

L'`URL` attendue est au format `https://go.readly.com/{folders}/{magazine_id}/{publication_id}` ou `https://go.readly.com/{folders}/{magazine_id}` voire `https://{pays}.readly.com/products/magazine/nom-du-magazine`. On peut aussi donner directement l'identifiant de la publication ou du magazine (24 caractères). 
Si c'est l'URL ou l'identifiant d'un magazine, c'est la publication la plus récente qui sera récupérée. 
Il est possible de mettre un chemin vers un fichier texte qui contient une liste d'URL. Toutes les lignes de ce fichier seront traitées. Il est possible de générer cette liste avec `readly_latest.py`. 

L'option `--token TOKEN` (optionnelle) permet de préciser le token d'API du compte Readly. On peut mettre en paramètre soit le token, soit un fichier qui contient le token. 
Si le token n'est pas renseigné, il est lu du fichier de config `auth_token`. 
Si le token n'est pas disponible dans ce fichier de config, il est demandé et sera sauvegardé dans `auth_token`. 

L'option `--output-folder "FOLDER"` ou `-o FOLDER` (optionnelle) permet de préciser le répertoire dans lequel seront stockés les fichiers. 
Si l'option n'est pas renseignée, le répertoire `DOWNLOADS` sera utilisé. 

L'option `--pattern "PATTERN"` (optionnelle) permet de choisir la façon dont les fichiers seront nommés. 
Les champs disponibles sont : `title` (le titre du magazine), `issue` (le nom de ce numéro), `date` (la date de publication). 
Si l'option n'est pas renseignée, le pattern utilisé sera `"title - issue (date)"`. 

L'option `--image-format {jpeg|webp}` ou `-i {jpeg|webp}` (optionnelle) permet de choisir le format des images qui seront utilisées. 
Les valeurs possibles sont : `jpeg` et `webp`. 
Si l'option n'est pas renseignées, le format `jpeg` sera utilisé. 
 
L'option `--quality QUALITY` ou `-q QUALITY` (optionnelle) permet de choisir la qualité de l'image. 
La valeur attendue doit être en `100` (meilleure qualité) et `0` (pire qualité). 
Si l'option n'est pas renseignées, la qualité `70` sera utilisé. 

L'option `--container-format {cbz|pdf}` ou `-c {cbz|pdf}` (optionnelle) permet de choisir le format du fichier dans lequel seront regroupées les images. 
Les valeurs possibles sont : `cbz` et `pdf`. 
Si l'option n'est pas renseignées, le format `pdf` sera utilisé. 

L'option `--low-quality` permet de récupérer les images en basse qualité. Attention, elles ne sont pas toujours disponibles. 

L'option `--dpi DPI` (optionnelle) permet de choisir le DPI des images enregistrées. 
Si l'option n'est pas renseignées, le DPI original des images sera conservé. 

L'option `--user-agent "USERAGENT"` (optionnelle) permet de choisir un user-agent spécifique à utiliser. 
Si l'option n'est pas renseignées, le user-agent `okhttp/3.12.1` sera utilisé. 

L'option `--pause SECONDS` ou `-p SECONDS` (optionnelle) permet de définir une pause (en secondes) à respecter entre chaque fichier récupéré. 
Si l'option n'est pas renseignées, aucune pause (`0`) ne sera appliquée. 

L'option `--max-dl` permet de définir combien de publications seront téléchargées au maximum dans le cas où l'URL correspond à une série. 
Si l'option n'est pas renseignée, une seule publication sera téléchargée (la plus récente). 

L'option `--no-clean` (optionnelle) permet de ne pas supprimer le répertoire temporaire dans lequel les images sont sauvegardées. 
Si l'option n'est pas renseignées, le répertoire temporaire sera supprimé après création du fichier CBZ ou PDF. 

L'option `--get-articles` (optionnelle) permet de dire que l'on souhaite télécharger aussi les articles qui sont parfois attachés à des publications. Attention à bien utiliser l'option `--no-clean`, sinon les articles seront supprimés en même temps que le répertoire temporaire. 
Si l'option n'est pas renseignées, les articles ne sont pas téléchargés. 

L'option `--get-articles-only` (optionnelle) permet de dire que l'on souhaite télécharger uniquement les articles qui sont parfois attachés à des publications. L'utilisation de cette option implique que les images ne seront pas téléchargées et le répertoire temporaire ne sera pas supprimé. Aucun fichier CBZ ou PDF ne sera généré.  

L'option `--create-token` (optionnelle) permet de faire une demande de nouveau token. L'utilisation de cette option implique qu'aucune autre action ne sera effectuée. 

L'option `--version` (optionnelle) permet d'afficher la version actuelle de l'application et de la comparer à la dernière version disponible. 
  
  
### Exemples 
#### Télécharger un magazine
```
python readly_get.py https://go.readly.com/magazines/category/it-technology/5e1f18a6d9e840630147fd25/60267250adeadd000d8c86e6
```
qui est l'équivalent de 
```
python readly_get.py --token auth_token --output-folder DOWNLOADS --pattern "title - issue (date)" --image-format jpeg --quality 70 --container-format pdf https://go.readly.com/magazines/category/it-technology/5e1f18a6d9e840630147fd25/60267250adeadd000d8c86e6
```
Le script va lire le token dans le fichier `auth_token`. Il va ensuite récupérer toutes les pages du magazine au format JPEG (qualité 70), puis compiler un PDF qui aura comme nom `Python & C++ for Beginners - February 2021 (2021-02-18).pdf`. 
Le répertoire temporaire sera supprimé. 

#### Télécharger le dernier numéro d'un magazine
```
python readly_get.py https://go.readly.com/magazines/category/it-technology/5e1f18a6d9e840630147fd25
```
Dans cet exemple, l'identifiant de la publication n'est pas mentionné, seul l'identifiant du magazine l'est. 
Le script va regarder quelle est la dernière publication disponible pour ce magazine et va la télécharger. 

#### Télécharger les articles d'un magazine
```
python readly_get.py --get-articles-only https://go.readly.com/magazines/category/news_politics/55e012198ea57fe8d300002e
```
Le script va récupérer uniquement les articles liés au magazine. 

#### Télécharger un magazine en CBZ qui utilise des images WEBP
```
python readly_get.py --image-format webp --quality 70 --container-format cbz https://go.readly.com/magazines/category/comics/5326c3fd01704d0ca7000034
```
L'avantage du format d'image WEBP est qu'il permet d'avoir une meilleure qualité d'image pour une taille inférieure à une image JPEG équivalente. 
Malheureusement, le format PDF ne permet pas d'inclure ce format d'image. Il faut donc utiliser le format `CBZ` en tant que container. 
  
   
   
   
## readly_latest.py
Permet de lister les publications Readly les plus récentes. 
 
### Utilisation
```
usage: readly_latest.py [-h] [--type TYPE] [--limit NUMBER] [--countries COUNTRIES] [--languages LANGUAGES]
                        [--categories CATEGORIES] [--output OUTPUT_FILE]

Script to display Readly latest publications.

optional arguments:
  -h, --help            show this help message and exit
  --type TYPE, -t TYPE  Type of publications ("magazines", "newspapers" or "magazines,newspapers").
                        Default="magazines,newspapers".
  --limit NUMBER, -i NUMBER
                        Number of issues per type. Default=25
  --countries COUNTRIES, -c COUNTRIES
                        Coutries of publications (2-letters format) coma separated. Default="" (all coutries).
  --languages LANGUAGES, -l LANGUAGES
                        Languages of publications (2-letters format) coma separated. Default="" (all languages).
  --categories CATEGORIES, -a CATEGORIES
                        Categories of publications coma separated. Default="" (all categories).
  --output OUTPUT_FILE, -o OUTPUT_FILE
                        Output to a file.
```

L'option `--type TYPE` ou `-t TYPE` (optionnelle) permet préciser de le type de publications recherchées. 
Les valeurs possibles sont : `magazines`, `newspapers` ou `"magazines,newspapers"`.
Si l'option n'est pas renseignées, le types utilisé sera `"magazines,newspapers"`. 

L'option `--limit NUMBER` ou `-i NUMBER` (optionnelle) permet de limiter le nombre de publications à retourner. 
Si l'option n'est pas renseignées, `25` publications (par type) seront retournées. 

L'option `--countries COUNTRIES` ou `-c COUNTRIES` (optionnelle) permet de limiter les pays d'origine des publications. Il faut indiquer les codes pays (2 caractères). 
On peut mettre plusieurs pays, séparés par des virgules. Par exemple pour filtrer sur tous les pays anglophones : `--countries "US,GB,IE,AU,NZ,CA"`.
Si l'option n'est pas renseignées, aucun filtre de pays ne sera appliqué. 

L'option `--languages LANGUAGES` ou `-l LANGUAGES` (optionnelle) permet de limiter les langues d'origine des publications. Il faut indiquer les codes langue (2 caractères). 
On peut mettre plusieurs langues, séparées par des virgules. Par exemple pour filtrer sur les publications en anglais ou en français : `--languages "en,fr"`.
Si l'option n'est pas renseignées, aucun filtre de langue ne sera appliqué. 

L'option `--categories CATEGORIES` ou `-a CATEGORIES` (optionnelle) permet de limiter les catégories des publications. 
Si l'option n'est pas renseignées, aucun filtre de catégorie ne sera appliqué. 

L'option `--output OUTPUT_FILE` ou `-o OUTPUT_FILE` (optionnelle) permet d'enregistrer le résultat de la requète dans le fichier `OUTPUT_FILE`. 
Ce fichier pourra ensuite être utilisé en entrée de `readly_get.py`. 
Si l'option n'est pas renseignées, le résultat sera affiché sur la sortie standard. 

### Exemples 
#### Lister les magazines en langue anglaise et les enregistrer dans un fichier
```
python readly_latest.py --languages en --output liste.txt
```

## Installation 
### Prérequis
- [Python 3.9+](https://www.python.org/downloads/windows/) (non testé avec les versions précédentes)
- pip

### Windows
- En ligne de commande, on clone le repo : 
```
git clone https://github.com/izneo-get/readly-get.git
cd readly-get
```
- (optionnel) On crée un environnement virtuel Python dédié : 
```
python -m venv env
env\Scripts\activate
```
- On installe les dépendances : 
```
python -m pip install --upgrade pip
pip install -r requirements.txt
```
- On peut exécuter le script :
```
python readly_get.py
```
- (optionnel) On quitte le l'environnement virtuel : 
```
deactivate
```

Si vous avez une erreur à cause d'une librairie SSL manquante, vous pouvez essayer de l'installer avec la commande :  
```
pip install pyopenssl
```
Si cela ne fonctionne pas, vous pouvez télécharger [OpenSSL pour Windows](http://gnuwin32.sourceforge.net/packages/openssl.htm). 


changelog.md
# Changelog

## readly_get.py

### Version 01.05 (2022-08-05)
- [CHANGE] Quand on précise l'URL d'une série, on peut télécharger plusieurs publications. 
- [NEW] Nouveau paramètre `--max-dl` qui permet de dire combien de publications on veut télécharger dans une série. 

### Version 01.04 (2021-03-23)
- [NEW] Possibilité de récupérer les images originales en low quality avec l'option `--default-quality` (pas disponible sur toutes les publications). 
- [NEW] Sauvegarde du token avec l'option `--create-token`. 
- [CHANGE] Qaulité par défaut est désormais `85` (au lieu de `70`). 

### Version 01.03 (2021-03-10)
- [NEW] Possibilité de mettre en entrée un fichier contenant une liste d'URLs. 
- [CHANGE] DPI par défaut est désormais `300` (au lieu de "celui de l'image d'origine"). 
 
### Version 01.02 (2021-03-09)
- [CHANGE] Plus de souplesse sur les URL acceptée. 
- [CHANGE] Possibilité d'indiquer directement l'identifiant de la publication.

### Version 01.01 (2021-03-09)
- [NEW] Ajout de l'option `--dpi DPI` qui permet de choisir le DPI des images. 
- [CHANGE] Amélioration de la procédure de vérification des tokens. 
- [CHANGE] Plus de souplesse dans les URL acceptées. 

### Version 01.00 (2021-03-08)
- [NEW] Première version. 
- [NEW] Permet de récupérer un magazine sur Readly.

## readly_latest.py 

### Version 01.00 (2021-03-10)
- [NEW] Version initiale. 


readly_latest.py
# -*- coding: utf-8 -*-
__version__ = "01.00"

import requests
import json
import sys
import argparse

if __name__ == "__main__":
    # Parse des arguments passés en ligne de commande.
    parser = argparse.ArgumentParser(
        description="""Script to display Readly latest publications."""
    )
    parser.add_argument(
        "--type",
        "-t",
        type=str,
        default="magazines,newspapers",
        help='Type of publications ("magazines", "newspapers" or "magazines,newspapers"). Default="magazines,newspapers".',
    )
    parser.add_argument(
        "--limit",
        "-i",
        metavar="NUMBER",
        type=int,
        default=25,
        help="Number of issues per type. Default=25",
    )
    parser.add_argument(
        "--countries",
        "-c",
        type=str,
        default="",
        help='Coutries of publications (2-letters format) coma separated. Default="" (all coutries).',
    )
    parser.add_argument(
        "--languages",
        "-l",
        type=str,
        default="",
        help='Languages of publications (2-letters format) coma separated. Default="" (all languages).',
    )
    parser.add_argument(
        "--categories",
        "-a",
        type=str,
        default="",
        help='Categories of publications coma separated. Default="" (all categories).',
    )
    parser.add_argument(
        "--output",
        "-o",
        type=str,
        metavar="OUTPUT_FILE",
        default="",
        help="Output to a file.",
    )

    args = parser.parse_args()
    pub_type = args.type
    per_page = args.limit
    origin = ""
    countries = args.countries  # FR%2CUS%2CGB%2CIE%2CAU%2CNZ%2CCA
    countries = countries.replace(",", "%2C").upper()
    languages = args.languages  # en%2Cfr
    languages = languages.replace(",", "%2C")
    categories = args.categories  #
    categories = categories.replace(",", "%2C")
    output_file = args.output

    fo = None
    if output_file:
        fo = open(output_file, "w", encoding="utf-8")
    for p_type in pub_type.split(","):
        print(f"# {p_type.upper()}", file=fo)
        url = f"https://d3og6tlt23zks5.cloudfront.net/{p_type}?ppage=1&per_page={per_page}&origin={origin}&countries={countries}&languages={languages}&categories={categories}"

        res = requests.get(url)
        if res.status_code != 200:
            print("[ERROR] Can't find latest.")
            sys.exit()

        issues = json.loads(res.text)

        for issue in issues["content"]:
            issue["date"] = issue["publish_date"][: len("YYYY-MM-DD")]
            if "issue" not in issue:
                issue["issue"] = issue["date"]

            print(f"# {issue['title']} - {issue['issue']} ({issue['date']})", file=fo)
            print(f"{issue['id']}", file=fo)
        print("", file=fo)


readly_get.py
# -*- coding: utf-8 -*-
__version__ = "01.05"
"""
Source : https://github.com/izneo-get/readly-get

URL list: 
https://api.readly.com/urls

"""

import requests
import sys
import os
import re
import argparse
import readly


def clean_name(name):
    """Permet de supprimer les caractères interdits dans les chemins.

    Parameters
    ----------
    name : str
        La chaine de caractère d'entrée.

    Returns
    -------
    str
        La chaine purgée des tous les caractères non désirés.
    """
    chars = '\\/:*<>?"|'
    for c in chars:
        name = name.replace(c, "_")
    name = re.sub(r"\s+", " ", name)
    name = re.sub(r"\.+$", "", name)
    return name


def check_version():
    latest_version_url = "https://raw.githubusercontent.com/izneo-get/readly-get/main/VERSION"
    res = requests.get(latest_version_url)
    if res.status_code != 200:
        print(f"Version {__version__} (impossible to check official version)")
    else:
        latest_version = res.text.strip()
        if latest_version == __version__:
            print(f"Version {__version__} (official version)")
        else:
            print(f"Version {__version__} (official version is different: {latest_version})")
            print("Please check https://github.com/izneo-get/readly-get/releases/latest")
    print()


def is_valid_url(url):
    return (
        re.match("https://go.readly.com(.*)/(.+?)/(.+)", url)
        or re.match("https://(.+?).readly.com/products/(.+)", url)
        or re.match("([\d\w]{24})", url)
    )


if __name__ == "__main__":
    # Parse des arguments passés en ligne de commande.
    parser = argparse.ArgumentParser(description="""Script to save a Readly publication.""")
    parser.add_argument(
        "url",
        type=str,
        default="",
        nargs="?",
        help="URL of the desired publication or publication_id.",
    )
    parser.add_argument(
        "--token",
        type=str,
        default="auth_token",
        help='Authentication token or file containing the token. Default="auth_token".',
    )
    parser.add_argument(
        "--output-folder",
        "-o",
        type=str,
        default="DOWNLOADS",
        help='Folder to save your publication. Default="DOWNLOADS".',
    )
    parser.add_argument(
        "--pattern",
        type=str,
        default="title - issue (date)",
        help='Pattern to name your file (available: "title", "issue", "date"). Default="title - issue (date)".',
    )
    parser.add_argument(
        "--image-format",
        "-i",
        type=str,
        choices=["jpeg", "webp"],
        default="jpeg",
        help='Image format saved (available: "jpeg", "webp"). Default="jpeg".',
    )
    parser.add_argument(
        "--quality",
        "-q",
        type=int,
        default=85,
        help='Image quality (100 = best quality). Default="85".',
    )
    parser.add_argument(
        "--container-format",
        "-c",
        type=str,
        choices=["pdf", "cbz"],
        default="pdf",
        help='Output file type (available: "cbz", "pdf"). Default="pdf".',
    )
    parser.add_argument(
        "--low-quality",
        action="store_true",
        default=False,
        help="Get default low quality images instead of HQ images.",
    )
    parser.add_argument(
        "--dpi",
        type=int,
        default=300,
        help='Image DPI (0 = original DPI). Default="300".',
    )
    parser.add_argument("--user-agent", type=str, default=None, help="User-agent to use.")
    parser.add_argument(
        "--pause",
        "-p",
        type=float,
        metavar="SECONDS",
        default=0,
        help='Make a pause (in seconds) between two pages. Default="0"',
    )
    parser.add_argument(
        "--max-dl",
        type=int,
        default=1,
        help='Max number of issues to download. Default="1".',
    )
    parser.add_argument(
        "--no-clean",
        action="store_true",
        default=False,
        help="Don't delete the temp folder where the images are stored.",
    )
    parser.add_argument(
        "--get-articles",
        action="store_true",
        default=False,
        help='Also download attached articles. Use with "--no-clean" option, or files will be deleted.',
    )
    parser.add_argument(
        "--get-articles-only",
        action="store_true",
        default=False,
        help='Download only attached articles (no image). Won\'t create PDF / CBZ file. Will force "--no-clean" option.',
    )
    parser.add_argument(
        "--create-token",
        action="store_true",
        default=False,
        help="Create a new token.",
    )
    parser.add_argument(
        "--version",
        action="store_true",
        default=False,
        help="Current version.",
    )

    args = parser.parse_args()
    url = args.url
    auth_token = args.token
    output_folder = args.output_folder
    output_pattern = args.pattern
    image_format = args.image_format
    quality = args.quality
    container_format = args.container_format
    use_default = args.low_quality
    dpi = args.dpi
    pause_sec = args.pause
    no_clean = args.no_clean
    version = args.version
    create_token = args.create_token
    max_dl = args.max_dl
    get_articles = args.get_articles
    get_articles_only = args.get_articles_only

    if create_token:
        new_token = readly.Readly.create_token()
        if new_token:
            print(f"{new_token}")
            answer = "?"
            while answer.upper() not in ("Y", "N", "O", "Q", ""):
                answer = input("Save token? [Y]es (default) / [N]o: ")
            if answer.upper() == "Q":
                sys.exit()
            if answer.upper() in ("Y", "O", ""):
                auth_token_file = "auth_token"
                if os.path.exists(auth_token_file):
                    answer = "?"
                    while answer.upper() not in ("Y", "N", "O", "Q", ""):
                        answer = input("File already exists. Overwrite? [Y]es / [N]o (default): ")
                    if answer.upper() == "Q":
                        sys.exit()
                else:
                    answer = "Y"
                if answer.upper() in ("Y", "O"):
                    with open(auth_token_file, "w") as f:
                        f.write(new_token)
                    print(f'[INFO] Token saved in "{auth_token_file}".')
        else:
            print("[ERROR] Impossible to create a new token...")
        sys.exit()

    check_version()
    if version:
        sys.exit()

    # Lecture du token.
    if os.path.exists(auth_token):
        auth_token = open(auth_token, "r").readline().strip()

    rdly = readly.Readly(auth_token)
    is_token_ok = rdly.is_token_ok()
    if not is_token_ok:
        print(f'[ERROR] Invalid token ("{auth_token}")...')
    while not is_token_ok:
        auth_token = input("Authentication token: ")
        if auth_token.upper() == "Q":
            sys.exit()
        if os.path.exists(auth_token):
            auth_token = open(auth_token, "r").readline()
        rdly = readly.Readly(auth_token)
        is_token_ok = rdly.is_token_ok()
        if not is_token_ok:
            print(f'[ERROR] Invalid token ("{auth_token}")...')
        else:
            answer = "?"
            while answer.upper() not in ("Y", "N", "O", "Q", ""):
                answer = input("Save token? [Y]es (default) / [N]o: ")
            if answer.upper() == "Q":
                sys.exit()
            if answer.upper() in ("Y", "O", ""):
                auth_token_file = "auth_token"
                if os.path.exists(auth_token_file):
                    answer = "?"
                    while answer.upper() not in ("Y", "N", "O", "Q", ""):
                        answer = input("File already exists. Overwrite? [Y]es / [N]o (default): ")
                    if answer.upper() == "Q":
                        sys.exit()
                else:
                    answer = "Y"
                if answer.upper() in ("Y", "O"):
                    with open(auth_token_file, "w") as f:
                        f.write(auth_token)
                    print(f'[INFO] Token saved in "{auth_token_file}".')

    rdly.output_folder = output_folder
    rdly.img_format = image_format
    rdly.img_quality = quality
    rdly.container_format = container_format
    rdly.pause_sec = pause_sec
    rdly.no_clean = no_clean
    rdly.get_articles = get_articles
    rdly.dpi = dpi
    rdly.use_default = use_default

    is_command_line = True

    # Lecture de l'URL.
    all_urls = []
    if not os.path.isfile(url):
        # C'est une URL ou un ID qui nous a été donné.
        while not is_valid_url(url):
            url = input('URL of publication or publication_id ("Q" to quit): ')
            is_command_line = False
            if url.upper() == "Q":
                sys.exit()
        all_urls.append(url)
    else:
        # C'est un fichier d'URLs qui nous a été donné.
        all_urls = [
            line.strip()
            for line in open(url, "r", encoding="utf-8").readlines()
            if len(line.strip()) > 0 and line.strip()[0] != "#"
        ]

    # On boucle sur toutes les URLs.
    for url in all_urls:
        print(f"[INFO] URL: {url}")
        # Id de publication
        match = re.match("([\d\w]{24})", url)
        if match:
            publication_id = match[1]
        else:
            # URL indirecte.
            if re.match("https://(.+?).readly.com/products/(.+)", url):
                res = requests.get(url)
                if res.status_code != 200:
                    print(f'[ERROR] Invalid URL "{url}".')
                    sys.exit()
                else:
                    match = re.search(r"\"publication_id\":\"([\d\w]+)\"", res.text)
                    if not match or not match[1]:
                        print(f'[ERROR] Invalid URL "{url}".')
                        sys.exit()
                    url = f"https://go.readly.com/magazines/{match[1]}"

            match = re.match("https://go.readly.com(.*)/(.+?)/(.+)", url)
            if not match:
                print(f'[ERROR] Invalid URL "{url}".')
                sys.exit()
            category, magazine_id, publication_id = match.groups()
            magazine_id = magazine_id.replace("/", "")
            publication_id = publication_id.replace("/", "")

        def download_issue(publication_id: str, infos: dict):
            print(
                f"[INFO] {publication_id} : \"{infos['title']} - {infos['issue']} ({infos['date']})\" will be downloaded."
            )
            output_filename = output_pattern
            for k in infos:
                output_filename = output_filename.replace(f"{k}", str(infos[k]))

            # Préparation du nom de sortie.
            output_filename = clean_name(output_filename)

            if get_articles_only:
                rdly.no_clean = True
                rdly.get_articles = True
                rdly.get_content = False
                print(f"[INFO] Articles only.")
            else:
                print(f"[INFO] Image format: {image_format.upper()}")
                print(f"[INFO] Image quality : {quality}")
                print(f"[INFO] Container format : {container_format.upper()}")
            rdly.download_publication(publication_id, save_as=output_filename)

        # Lecture des infos.
        infos = rdly.get_infos(publication_id)
        if not infos:
            # On n'a pas d'infos, c'est peut-être un ID de magazine.
            print(f'[WARNING] Invalid publication_id "{publication_id}"...')
            print(f'[INFO] Available publications for magazine_id "{publication_id}": ')
            publications = rdly.get_all_publications(publication_id)
            for p in publications:
                print(f"{p['id']}\t{p['title']} - {p['issue']} ({p['date']})")
            print(f"[INFO] {len(publications)} publications found.")

            if not is_command_line:
                res = "?"
                while not res.isdigit():
                    res = input(
                        f"How many issues do you want to download? (number or 'Q' to quit) (default: {max_dl}): "
                    )
                    if res.upper() == "Q":
                        sys.exit()
                    if not res:
                        res = str(max_dl)
                max_dl = int(res)
            for p in publications[:max_dl]:
                publication_id = p["id"]
                infos = rdly.get_infos(publication_id)
                if infos:
                    download_issue(publication_id, infos)
                else:
                    print(f'[ERROR] Invalid publication_id "{publication_id}"...')
        else:
            download_issue(publication_id, infos)
